echo: add echo server

Spiffyk opened the pull request:
This commit adds an echo server that reflects queries back at the sender. May be used for determining the baseline performance of QUIC (or, at least, of the `quic-go` implementation in combination with the client's QUIC implementation).

Spiffyk committed:
echo: add echo server

This commit adds an echo server that reflects queries back at the
sender. May be used for determining the baseline performance of QUIC
(or, at least, of the `quic-go` implementation in combination with the
client's QUIC implementation).

Spiffyk committed:
echo: replace io.ReadAll with stream.Read

ReadAll is apparently SLOW for our purposes. Reading into a fixed-size
buffer seems to be much better.

fcelda commented on cmd/echo/main.go:
```diff
@@ -0,0 +1,166 @@
+		return fmt.Errorf("read query: %w", err)
+	}
+
+	_, err = stream.Write(data[:n])
```
I think it would be useful if the echo server didn't blindly reflect any payload but at least checked that it has a DNS header with the QR bit cleared and also to set the QR bit in response.

fcelda commented on go.sum:
```diff
@@ -0,0 +1,141 @@
+github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
```
Please, do not include `go.sum`.

Spiffyk replied:
Removed and added to `.gitignore`.

fcelda commented on cmd/echo/main.go:
```diff
@@ -0,0 +1,166 @@
+		return fmt.Errorf("read query: %w", err)
+	}
+
+	_, err = stream.Write(data[:n])
```
I think it would be useful if the echo server didn't blindly reflect any payload but at least checked that it has a DNS header with the QR bit cleared and also to set the QR bit in response.

Spiffyk replied:
I had something like that in place at one point, but removed it - I think it is pretty useful to have even just the blind reflection, as that can be used for non-DNS-related testing.

Maybe a good compromise would be to have QR bit handling available as an option? If that would be acceptable, would you rather have it in "DNS-mode" or "non-DNS-mode" by default? I'm not really inclined either way :)

Spiffyk replied:
I added a `-dns` option (usable as `-dns=true` or `-dns=false`), defaulting to `true`. This makes the server parse the DNS query, check that the QR bit is cleared, and set the QR bit for the response, as suggested. When `-dns=false`, the server just blindly reflects the payload as it was originally.

fcelda commented on cmd/echo/main.go:
```diff
@@ -0,0 +1,166 @@
+	wg.Wait()
+}
+
+func handleStream(l log.Logger, stream quic.Stream) error {
```
This is a lot of copy-pasted code. Only the `handleStream` function is unique. I wonder if it could be abstracted into some sane interface.

Spiffyk replied:
Sounds good. I'm very new to Go, so I'll need to go study how that is done in the language, but hey, at least I'll get to learn something new today :grin:

Spiffyk replied:
After some tinkering and numerous frustrations with not knowing how to get modules to see each other, I managed to do some deduplication. Not sure how Go-idiomatic it is, I'm sure things could be improved. Suggestions are welcome.

Spiffyk replied:
I made some more improvements to the deduplication, using Go's generics for specific flags (i.e. `echo` has its `-dns` flag, and `proxy` has its `-backend` flag).

fcelda reviewed (changes requested):
Thank you for the contribution. Please, do you mind checking the comments below? I don't insist on all changes.

Spiffyk committed:
Exclude go.sum

Spiffyk committed:
server: code deduplication

Spiffyk committed:
Update quic-go, enable 0-RTT, more error logging

Spiffyk committed:
Restore the limit on maximum concurrent streams

Spiffyk committed:
echo: QR-bit-setting DNS mode

Spiffyk committed:
Refactor and deduplicate server parameters using generics

Spiffyk committed:
server: add option to specify TLS key log file

This allows developers to analyze the traffic using Wireshark, even when
it is encrypted.

Spiffyk committed:
server: random stream reset option (opt-in)

For testing purposes, a -reset option that randomly resets streams.
